api.bs

```diff
@@ -1702,26 +1702,36 @@ To perform <dfn>common matching logic</dfn>, given
 <div algorithm>
 To <dfn>fairly allocate credit</dfn>, given a [=list=] of doubles |credit| and an integer |value|:
 1.  Let |sumCredit| be the sum of |credit|'s [=list/items=].
-1.  Let |rawNormalizedCredit| be a new [=list=].
-1.  Let |normalizedCredit| be a new [=list=].
-1.  [=list/iterate|For each=] |item| of |credit|:
-      1. Let |rawNormalized| be |value| * |item| / |sumCredit|.
-      1. [=list/Append=] |rawNormalized| to |rawNormalizedCredit|.
-      1. Let |normalized| be |rawNormalized| rounded towards positive Infinity and converted to an integer.
-      1. [=list/Append=] |normalized| to |normalizedCredit|.
-1.  Let |shuffledFractionalIndices| be [=list/get the indices|the indices=] of |credit|, ordered randomly.
-1.  [=list/remove|Remove=] all indices |i| from |shuffledFractionalIndices| where |rawNormalizedCredit|[|i|] has no fractional part.
-1.  [=list/iterate|For each=] |index| of |shuffledFractionalIndices|:
-      1. If the sum of |normalizedCredit|'s [=list/items=] is equal to |value|, [=iteration/break=].
-      1. Decrement |normalizedCredit|[|index|] by 1.
-1.  Return |normalizedCredit|.
+1.  Let |unroundedCredit| be a new [=map=].
+1.  Let |roundedCredit| be a new [=map=].
+1.  [=list/iterate|For each=] |idx| of [=list/get the indices|the indices=] of |credit|:
+    1.  Let |rawNormalized| be |value| * |credit|[|idx|] / |sumCredit|.
+    1.  If |rawNormalizedCredit| is equal to |rawNormalizedCredit| rounded towards positive Infinity,
+        [=map/set=] |roundedCredit|[|idx|] to |rawNormalizedCredit|, converted to an integer.
+    1.  Otherwise, [=map/set|set=] |unroundedCredit|[idx] to |rawNormalized|.
+1.  While the [=map/size=] of |unroundedCredit| is greater than 1.
+    1.  Let |idx1| and |x1| be the [=map/key=] and [=map/value=] of the first [=map/entry=] of |unroundedCredit|.
+    1.  Let |idx2| and |x2| be the [=map/key=] and [=map/value=] of the second [=map/entry=] of |unroundedCredit|.
+    1.  Let |delta1| be min(ceiling(|x1|) - x1, x2 - floor(x2)).
+    1.  Let |delta2| be min(ceiling(|x2|) - x2, x1 - floor(x1)).
+    1.  Let |p| be |delta2| / (|delta1| + |delta2|)
+    1.  Let |r| be a random double between 0 and 1 (inclusive).
+    1.  If |r| is less than |p|, set |x1| to |x1| + |delta1|, and set |x2| to |x2 - |delta1|.
+    1.  Otherwise, set |x2| to |x2| + |delta2|, and set |x1| to |x1 - |delta2|.
```

Comment:
```suggestion
    1.  Let |delta1| be min(ceiling(|x1|) − |x1|, |x2| − floor(|x2|)).
    1.  Let |delta2| be min(ceiling(|x2|) − |x2|, |x1| − floor(|x1|)).
    1.  Let |p| be |delta2| / (|delta1| + |delta2|)
    1.  Let |r| be a random double between 0 and 1 (inclusive).
    1.  If |r| is less than |p|, set |x1| to |x1| + |delta1|, and set |x2| to |x2| − |delta1|.
    1.  Otherwise, set |x2| to |x2| + |delta2|, and set |x1| to |x1| − |delta2|.
```